schema-engine-tag/src/main/java/org/apache/iotdb/lsm/manager/FlushManager.java

```diff
@@ -0,0 +1,128 @@
+
+  private ScheduledExecutorService checkFlushThread;
+
+  private String flushDirPath;
+
+  private String flushFilePrefix;
+
+  private final int flushIntervalMs = 1000;
+
+  public FlushManager(
+      WALManager walManager, T memManager, String flushDirPath, String flushFilePrefix) {
+    this.walManager = walManager;
+    if (!(memManager instanceof IMemManager)) {
+      throw new ClassCastException("memManager must implement IMemManager");
+    }
+    this.memManager = (IMemManager) memManager;
+    this.flushDirPath = flushDirPath;
+    File flushDir = new File(flushDirPath);
+    flushDir.mkdirs();
+    this.flushFilePrefix = flushFilePrefix;
+    checkFlushThread = IoTDBThreadPoolFactory.newSingleThreadScheduledExecutor("LSM-Flush-Service");
+    ScheduledExecutorUtil.safelyScheduleAtFixedRate(
+        checkFlushThread,
+        this::checkFlush,
+        flushIntervalMs,
+        flushIntervalMs,
+        TimeUnit.MILLISECONDS);
+  }
+
+  public void checkFlush() {
+    synchronized (memManager) {
+      if (memManager.isNeedFlush()) {
+        List<R> flushRequests = memManager.getFlushRequests();
+        for (R flushRequest : flushRequests) {
+          flushRequest.setFlushDirPath(flushDirPath);
+          flushRequest.setFlushFileName(flushFilePrefix + "-0-" + flushRequest.getIndex());
+          flushRequest.setFlushDeleteFileName(
+              flushFilePrefix + "-delete" + "-0-" + flushRequest.getIndex());
+          flush(flushRequest);
+        }
+      }
+    }
+  }
+
+  private void updateWal(R request) {
+    int index = request.getIndex();
+    walManager.deleteWalFile(index);
+  }
+
+  private void renameFlushFile(T root, R request, FlushRequestContext context) {
+    TiFileOutputStream fileOutput = context.getFileOutput();
+    try {
+      fileOutput.close();
+      String flushFileName = request.getFlushFileName() + SchemaRegionConstant.TMP;
+      File flushFile = new File(this.flushDirPath, flushFileName);
+      File newFlushFile = new File(this.flushDirPath, request.getFlushFileName());
+      flushFile.renameTo(newFlushFile);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void flush(R flushRequest) {
+    FlushRequestContext flushRequestBaseContext = new FlushRequestContext();
+    process((T) flushRequest.getMemNode(), flushRequest, flushRequestBaseContext);
+  }
+
+  @Override
+  public void preProcess(T root, R request, FlushRequestContext context) {
+    String flushFileName = request.getFlushFileName() + SchemaRegionConstant.TMP;
+    File flushFile = new File(this.flushDirPath, flushFileName);
+    try {
+      if (!flushFile.exists()) {
+        flushFile.createNewFile();
+      }
+      context.setFileOutput(new TiFileOutputStream(flushFile));
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void postProcess(T root, R request, FlushRequestContext context) {
+    memManager.removeMemData(request);
```

Comment:
If there's exception occurred during deleting WAL and renaming tmp file, we will lose the memory data and disk data, thus we should delete memory data in last step.